### PR TITLE
[Experimental] Form modelling

### PR DESCRIPTION
**Pull request by davidrapson**
This is an exploration of modelling forms to answer some of the questions posed in https://github.com/biglotteryfund/blf-alpha/issues/295

forms

**Structure**

The forms are modelled in the main parts

- The router, which handles boilerplate code involved in rendering each step, displaying errors, and persisting progress
- Some plumbing code in `form-model` which helps us model individual forms 
- `example-form.js` which is the main business logic, defines a form as a series of steps with a schema for individual fields 

**Goals**

- Aims to validate if how easy it is to model multi-page forms with conditional fields without needing to hand-craft individual fields (i.e. can we model a form schema and add fields as requirements change over time)
- Should achieve comprehensive validation for each field
- Validates if the above can be done using server side code, with client-side code only use for progressive enhancement

**Caveats**

This change involves updating to the latest version of `express-validator` as it has a much more flexible API which enables this approach. This means the **tests will fail** as the materials form still depends on the old API.

### CODE REVIEW

File: assets/js/modules/forms.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    const formEl = $('.js-application-form');
+    if (formEl.length === 0) {
+        return;
+    }
+
+    const conditionalFields = $('.js-conditional-for');
+    conditionalFields.each(function() {
+        const $el = $(this);
+        const conditionalType = $el.data('conditionalType');
+        if (conditionalType === 'radio');
```
Comment (mattandrews) [inline assets/js/modules/forms.js]: Is this valid?!

File: assets/js/modules/forms.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    const conditionalFields = $el.data('conditionalFields');
+    $radios.on('change', function() {
+        const selectedValue = $(this).val();
+        conditionalFields.forEach(field => {
+            const $triggerField = $(`[data-conditional-field="${field.triggerField}"]`);
+            if (selectedValue === field.triggerOnValue) {
+                $triggerField.attr('hidden', false);
+            } else {
+                $triggerField.attr('hidden', true);
+            }
```
Comment (mattandrews) [inline assets/js/modules/forms.js]: You could rewrite this as:
```
$triggerField.attr("hidden", function() {
    return selectedValue === field.triggerOnValue;
});
```

File: controllers/forms/example-form.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                            .isEmpty()
+                            .withMessage('Last-name must be provided');
+                    }
+                },
+                {
+                    type: 'text',
+                    name: 'organisation',
+                    label: 'Organisation',
+                    isRequired: true,
+                    validator: function(field /*, reqBody */) {
```
Comment (mattandrews) [inline controllers/forms/example-form.js]: Do we need the comment?

File: controllers/forms/example-form.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                    }
+                },
+                {
+                    type: 'text',
+                    name: 'other-colour',
+                    label: 'Other Colour',
+                    conditionalOn: {
+                        name: 'colour',
+                        value: 'other',
+                        fallbackText: 'Only fill this in if you have selected ’other‘ for colour'
```
Comment (mattandrews) [inline controllers/forms/example-form.js]: crazy quotes!

File: controllers/forms/form-model.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
+const { find, flatMap, has, get } = require('lodash');
+
+function enhanceField(field, associatedConditionalFields) {
```
Comment (mattandrews) [inline controllers/forms/form-model.js]: This function (and the one below it) feels a bit magic, could you add a few comments?

File: controllers/forms/example-form.js (hunk 4 above)
Comment (davidrapson) [reply controllers/forms/example-form.js]: :trollface:

File: controllers/forms/example-form.js (hunk 3 above)
Comment (davidrapson) [reply controllers/forms/example-form.js]: No. It's also not true, so should go…

**Comment by davidrapson** [comment]
As discussed, if you could (re)review this tomorrow as I did a bunch more work I can get this merged in.

File: controllers/apply/forms/create-form-model.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        });
+
+        return enhancedFields;
+    };
+}
+
+/**
+ * Create a step based on a schema.
+ * Allows us to pass a relatively consise schema for the step,
+ * this funciton then adds some additional computed methods on top.
```
Comment (mattandrews) [inline controllers/apply/forms/create-form-model.js]: Typo in "funciton"

File: controllers/apply/forms/create-form-model.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+ * - withValues allows the current form data for a step to be passed in and the values attached to each field
+ * - getValidators collects all validators associated with each field for express-valdiator
```
Comment (mattandrews) [inline controllers/apply/forms/create-form-model.js]: Typo on "valdiator" (although you're in [good company](https://github.com/search?o=desc&q=valdiator&s=indexed&type=Code&utf8=%E2%9C%93)).

File: views/pages/experimental/apply/form.njk [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+{% macro inputText(field) %}
+    <label class="ff-label" for="field-{{ field.name }}">{{ field.label }}</label>
+    <input
+        class="ff-text"
+        type="text"
+        id="field-{{ field.name }}"
+        name="{{ field.name }}"
+        value="{% if field.value %}{{ field.value }}{% endif %}"
+        {# {% if field.isRequired %}required{% endif %} #}
```
Comment (mattandrews) [inline views/pages/experimental/apply/form.njk]: Could maybe add the `aria-required` attribute here too, but given this isn't intended to be user-facing it's probably not worth it for now.

File: views/pages/experimental/apply/review.njk [hunk 9]
```diff
@@ -0,0 +1,27 @@
+{% extends "../../../layouts/main.njk" %}
+
+{% block content %}
+    <div class="content-box inner inner--wide-only accent--pink a--border-top">
+        <div class="u-constrained-content">
+            <h1>Check Your Answers Before Sending Your Application</h1>
```
Comment (mattandrews) [inline views/pages/experimental/apply/review.njk]: BuzzFeed approves of your capitalisation choices 👍

**Pull request merged**